sno/pk_generation.py

```diff
@@ -0,0 +1,176 @@
+        If multiple ImportSources are supplied, wraps those that lack a primary key and returns a new list.
+        """
+        if isinstance(source_or_sources, Iterable):
+            sources = source_or_sources
+            return [cls.wrap_if_needed(s, repo) for s in sources]
+
+        source = source_or_sources
+        if not source.schema.pk_columns:
+            return PkGeneratingImportSource(source, repo)
+        else:
+            return source
+
+    def __init__(self, delegate, repo):
+        self.delegate = delegate
+        self.load_data_from_repo(repo)
+
+    def load_data_from_repo(self, repo):
+        tree = repo.head_tree
+        generated_pks_blob = None
+
+        if tree is not None:
+            try:
+                generated_pks_blob = tree / self.dest_path / self.GENERATED_PKS_PATH
+            except KeyError:
+                pass
+
+        if not generated_pks_blob:
+            self.pk_col = self.DEFAULT_PK_COL
+            self.primary_key = self.pk_col["name"]
+            self.pk_to_hash = {}
+            self.hash_to_pks = {}
+            self.hash_to_unused_pks = {}
+            self.next_new_pk = 1
+            return
+
+        data = json_unpack(generated_pks_blob.data)
+        self.pk_col = data["pkSchema"]
+        self.primary_key = self.pk_col["name"]
+
+        adapt_pk = int if self.pk_col["dataType"] == "integer" else str
+        self.pk_to_hash = {
+            adapt_pk(pk): feature_hash
+            for pk, feature_hash in data["generatedPks"].items()
+        }
+
+        self.hash_to_pks = self._invert_pk_map(self.pk_to_hash)
+        self.hash_to_unused_pks = self._invert_pk_map(self.pk_to_hash)
+
+        if self.hash_to_pks:
+            self.next_new_pk = max(self.pk_to_hash) + 1
+        else:
+            self.next_new_pk = 1
+
+    def encode_generated_pk_data(self, relative=False):
+        path = self.GENERATED_PKS_PATH
+        if not relative:
+            path = "/".join(self.dest_path, self.GENERATED_PKS_PATH)
+
+        data = {
+            "pkSchema": self.pk_col,
+            "generatedPks": self.pk_to_hash,
+        }
+
+        return path, json_pack(data)
+
+    def _invert_pk_map(self, pk_to_hash):
+        result = {}
+        for pk, h in pk_to_hash.items():
+            result.setdefault(h, [])
+            result[h].append(pk)
+        return result
+
+    @property
+    @functools.lru_cache(maxsize=1)
+    def schema(self):
+        cols = self.delegate.schema.to_column_dicts()
+        return Schema.from_column_dicts([self.pk_col] + cols)
+
+    def features(self):
+        schema = self.schema
+        for feature in self.delegate.features():
+            feature[self.primary_key] = None
+            feature_hash = schema.hash_feature(feature)
+            feature[self.primary_key] = self.generate_pk(feature_hash)
+            yield feature
+
+    def generate_pk(self, feature_hash):
+        unused_pks = self.hash_to_unused_pks.get(feature_hash)
```

Comment:
Can you explain what/how `unused_pks` work? IMO we shouldn't be re-using numbers at all.

* Feature A gets imported
* and later deleted.
* then if an identical feature later gets imported again, it should get a new PK from the sequence.

Reply:
> Can you explain what/how unused_pks work?

Unused PKs are PKs that haven't yet been assigned during this import. I'll rename them to `unassigned_pks`.

Say you import 6 features, but some are identical: [A, A, A, B, B, C]
Say these are assigned primary keys: {A: [1,2,3], B: [4, 5], C: [6]}
This dict is called `hash_to_pks` and it exists at the end of the import.
Next time you import, you have a similar but different list of features: [A, A, B, B, D]
The same `hash_to_pks` dict is loaded again, but also duplicated as `hash_to_unused_pks`.

As you import the first A, you check the hash_to_unused_pks to see if you can find a primary key for it. `1` is the correct answer. But now, the 1 is removed from the list of unused_pks so that the next A encountered will be assigned `2`.

> IMO we shouldn't be re-using numbers at all.
The purpose of this dict - and of pk-less imports in general - is so that we DO want to reuse numbers. PK-less imports would be simpler if we could assign random PKs to every feature we encounter, but we want to assign a certain feature the same PK every time we reimport it. But then assign different PK if we see an identical feature a moment later. That's what this dict is for - keeping track of which PKs are the correct choice for feature A, but not using any of them more than once.

But you have another concern also:

> Feature A gets imported
> and later deleted.
> then if an identical feature later gets imported again, it should get a new PK from the sequence.

This is a non-goal of PK-less imports. Any local edits made should have no effect on how we re-import from the source next time we do an import (excepting that the user may find a way to edit the stored metadata). If local edits to features did have an effect on how we reimported from the same source, this would be inconsistent with current imports. Currently imports are stateless and depend only on the source, not on the local repo.

This PR makes imports stateful if they are PK-less imports, but only in-so-far as it tries to give the same features the same primary keys as they were given last time they were imported. It still doesn't let local feature edits affect the import.

Note that for regular imports using existing code - if you import the feature with PK=1, then delete it, then reimport from the same source, it reappears with the same PK as last time. So, PK-less imports is consistent with this.

The main use-case for this is mirroring, where there should generally be no local edits anyway.

Reply:
Cool, thanks for clearing that up :)